framework/src/task_manager/task_manager_core.c

```diff
@@ -591,59 +605,173 @@ int taskmgr_get_handle_by_pid(int pid)
+			if (broadcast_info == NULL) {
 				continue;
 			}
 			ret = ioctl(fd, TMIOC_BROADCAST, TM_PID(handle));
 			if (ret != OK) {
 				continue;
 			}
+			msg_broad.sival_ptr = broadcast_info;
 			(void)sigqueue(TM_PID(handle), SIGTM_BROADCAST, msg_broad);
 		}
 	}
 }
 
+static void taskmgr_broadcast_msg_init(void)
+{
+	int chk_idx;
+	int msg_num;
+	int broadcast_max_num;
+
+#ifdef CONFIG_USER_DEFINED_BROADCAST
+	broadcast_max_num = TM_BROADCAST_USER_DEFINE_MAX;
+#else
+	broadcast_max_num = TM_BROADCAST_DEFINE_MAX;
+#endif
+	for (chk_idx = 0; chk_idx < broadcast_max_num + CONFIG_TASK_MANAGER_MAX_TASKS; chk_idx++) {
+		msg_num = chk_idx + 1;
+		if (chk_idx < (broadcast_max_num - 1)) {
+			tm_broadcast_msg[chk_idx] = msg_num;
+		} else {
+			tm_broadcast_msg[chk_idx] = -1;
+		}
+	}
+}
+
 static int taskmgr_set_msg_cb(int type, void *data, int pid)
 {
 	int handle;
+	int ret;
+	int broadcast_max_num;
 
 	if (data == NULL) {
 		return TM_INVALID_PARAM;
 	}
-
 	handle = taskmgr_get_handle_by_pid(pid);
 	if (handle == TM_UNREGISTERED_APP) {
 		return handle;
 	}
-
 	if (type == TYPE_UNICAST) {
-		TM_UNICAST_CB(handle) = (_tm_unicast_t)data;
+		TM_UNICAST_CB(handle) = (_tm_cb_t)data;
 	} else {
-		TM_MSG_MASK(handle) = ((tm_broadcast_t *)data)->msg_mask;
-		TM_BROADCAST_CB(handle) = ((tm_broadcast_t *)data)->cb;
+#ifdef CONFIG_USER_DEFINED_BROADCAST
+	broadcast_max_num = TM_BROADCAST_USER_DEFINE_MAX;
+#else
+	broadcast_max_num = TM_BROADCAST_DEFINE_MAX;
+#endif
+		if (!((tm_broadcast_msg[0] == 1) || (tm_broadcast_msg[0] == -1))) {
+			taskmgr_broadcast_msg_init();
+		}
+		ret = taskmgr_check_broad_msg(((tm_broadcast_t *)data)->msg);
+		if (ret == TM_UNREGISTERED_MSG) {
+			return ret;
+		}
+		tm_broadcast_t *broadcast_info = (tm_broadcast_t *)TM_ALLOC(sizeof(tm_broadcast_t));
+		((tm_broadcast_t *)broadcast_info)->flink = NULL;
+		((tm_broadcast_t *)broadcast_info)->msg = ((tm_broadcast_t *)data)->msg;
+		((tm_broadcast_t *)broadcast_info)->cb = ((tm_broadcast_t *)data)->cb;
+		((tm_broadcast_t *)broadcast_info)->cb_data = ((tm_broadcast_t *)data)->cb_data;
+		sq_addlast((FAR sq_entry_t *)broadcast_info, &TM_BROADCAST_INFO_LIST(handle));
 	}
 	return OK;
 }
 
+static int taskmgr_get_new_broadcast_msg(void)
+{
+	int chk_idx;
+	int msg_num;
+	int broadcast_max_num;
+
+#ifdef CONFIG_USER_DEFINED_BROADCAST
+	broadcast_max_num = TM_BROADCAST_USER_DEFINE_MAX;
+#else
+	broadcast_max_num = TM_BROADCAST_DEFINE_MAX;
+#endif
+	if (!((tm_broadcast_msg[0] == 1) || (tm_broadcast_msg[0] == -1))) {
+		taskmgr_broadcast_msg_init();
+	}
+	for (chk_idx = 0; chk_idx < broadcast_max_num + CONFIG_TASK_MANAGER_MAX_TASKS; chk_idx++) {
+		msg_num = chk_idx + 1;
+		if (tm_broadcast_msg[chk_idx] == -1) {
+				tm_broadcast_msg[chk_idx] = msg_num;
+				return msg_num;
+		}
+	}
+	return TM_OUT_OF_MEMORY;
+}
+
+static int taskmgr_unset_broadcast_cb(int msg, int pid)
+{
+	int handle;
+	tm_broadcast_t *broadcast_info;
+
+	handle = taskmgr_get_handle_by_pid(pid);
```

Comment:
taskmgr_get_handle_by_pid returns handler id based on pid, but
it returns error number??
TM_UNREGISTERED_APP is one of tm_result_error_e, you have to redesign taskmgr_get_handle_by_pid.
tm_result_error_e taskmgr_get_handle_by_pid (int *res, int pid)
or 
you can use #define TM_ERROR_HANDLE_ID -1 and return it.
What I am meaning is that taskmgr_get_handle_by_pid  should return only pid or error type.

Reply:
Thank you for this comment, this issue will be solved at another commit, because taskmgr_get_handle_by_pid is called in many places.